### PR TITLE
Revise Architectural Tenets

### PR DESCRIPTION
**Pull request by snickell**
# TL;DR from the intro to [ARCHITECTURE.md](https://github.com/code-dot-org/code-dot-org/blob/ARCHITECTURE.md/ARCHITECTURE.md#architectural-tenets)

> **TL;DR**: work done at code.org should live in the code-dot-org monorepo, be deployed in the container as part of one of our existing monoliths, be integration tested by our central CI system prior to PR merge, use REST for client/server communication, and be implemented using Ruby/Rails/JS/TS/React/MySQL.

That's all the architectural tenets in one long mouhtful!

The rest of the doc is just going into more DETAIL on these tenets (including non obvious implications of the tenet), explaining WHY the tenet was chosen, and explaining any EXCEPTIONS to the tenet.

## Summary of changes relative to [previous Architectural Tenets google doc](https://docs.google.com/document/d/1esRqj082XEx77YiffA6vfbiQaFB5nYdazp9AU5cOtYI/edit?tab=t.0) from which this document was derived:
- Sorted tenets from more general to more specific, and thus renumbered them for the long-haul.
- Partially or wholly rewrote most Why and Exception sections to give better long-term motivation, and to focus exceptions on "valid exceptions going forward" vs mixing with "existing violations that are grandfathered in".
- Previous doc was oriented toward one to two year planning, not being a permanent-life document. Thus I removed TODO type plans (with the exception of containers) in favor of highlighting the tenets themselves.
- Includes "freshness and enforcement" clauses to encourage these to be read and taken seriously, ammended if needed, and deleted if no longer relevant.

I dropped three more specific tenets and replaced them with three more general tenets:
- Dropped tenet "Server-side rendering (vs. SPAs) for our teacher and partner tools" after consulting with Dave. This one was not in line with reality.
- Dropped tenet "Lightweight SPAs for our student activities" after consulting with Emily. This tenet wasn't wrong, but didn't really add anything at the current state. A few of its most salient points were shifted to the related "use REST" tenet.
- Dropped tenet "Use 3rd party auth providers", a good technical plan, but wasn't really a long term "one of 8 core tenets for the codebase".
- Added tenet "Continuous Integration of All Parts (vs. only decoupled unit based testing)"
- Added tenet “The Rails Way (vs. overfitting to today’s needs)"
- Added tenet "Managed Services that implement an Open Source Protocol / API (vs proprietary managed services or self-hosted open source services)"

For review purposes/reading, it may be easier to [view ARCHITECTURE.md as rendered markdown](https://github.com/code-dot-org/code-dot-org/blob/ARCHITECTURE.md/ARCHITECTURE.md#architectural-tenets).

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+Examples:
+
+| Collective Benefit Tomorrow | Trade Off Today |
+| :---- | :---- |
+| **“If my PR passes  CI it won’t break code.org”** | “CI is slow” |
+| **“I  can implement features/refactors that require changes across the codebase: even in areas I didn’t know about before (even if nobody knows about them anymore)”** | “I have to work inside the monorepo even if it’d be so much faster to deliver this quarter’s feature goals by spinning out my own repo without the overhead” |
+| **“After mastering a few key technologies I can work in almost any part of the codebase”** | “Even if hotnewframework is the very best option in 2120, it might be best if I use existingokframework from 2115 because we already use it widely and it gets the job done just fine.” |
+
+Historical violations of almost all tenets exist, and are de facto grandfathered in. However, these should not be interpreted as an invitation to follow suit. A strong majority of historical code already adheres to all the tenets, which is part of why they were picked in the first place.
```
Comment (cat5inthecradle) [inline ARCHITECTURE.md]:
```suggestion
Historical violations of almost all tenets exist, and are not a priority to fix. However, these should not be interpreted as an invitation to follow suit. A strong majority of historical code already adheres to all the tenets, which is part of why they were picked in the first place.
```

rewriting to avoid use of a loaded term

**Comment by cat5inthecradle** [comment]
I like the cleanup and it's use of more established examples for exceptions, and I love that we're putting this in the repo.

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+### Tenet 7: Use REST (vs. GraphQL) for APIs
+
+As we add APIs for our platform, we will publish them using REST. Client-side apps should communicate with the server using REST endpoints.
+
+<details>
+  <summary>Why?</summary>
+
+  * KISS: REST has been a simple and an effective match to our needs.  
+  * Our labs and other fat client-side JS apps don’t tend to have the “complex aggregation of dynamic data from many sources” property that benefits from more complex systems like GraphQL.  
+  * Memoization is easily applied to REST-based calls (good for hosting cost and at-scale events like hour of code).
```
Comment (davidsbailey) [inline ARCHITECTURE.md]: it feels like cacheability should be on this list -- REST GET requests are cacheable by cloudfront / firewalls / browsers while GraphQL requests generally are not. one possible phrasing:

```
REST APIs inherently support HTTP caching mechanisms (e.g., ETag, Last-Modified, Cache-Control), enabling better performance and reducing load on the server.
```

separately, I'm not sure what you mean by memoization -- if you do mean caching, I think we should call it caching or cacheability. if not, can you say more about what memoization means in this case?

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+### ⬆️ Above ⬆️ this line live the architecutral tenets. They should be kept at the top of the file and edited sparingly.
+
+### ⬇️ Below ⬇️ this line is general architecture information: please edit freely!
+
+---
+
+# Code.org's Overall Architecture...
+
+Overview of our architecture goes here, please fill me in!
```
Comment (davidsbailey) [inline ARCHITECTURE.md:218]: I'm lukewarm on including this because it feels like could quickly become a dumping ground. I'm guessing your goal is to keep non-tenets out of the tenets section above, which I support, but would like to consider some alternatives.

for example: perhaps just say "further documentation of current architecture can be found in Google Drive > Shared drives > Engineering > Architecture"?

**Review by davidsbailey — COMMENTED** [review]
this looks great Seth! The supporting evidence for each tenet looks great and the collapse/expand seems like a great mechanism for keeping the reader focused on the actual tenets. There are a couple issues below I'd like to discuss before we proceed.

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+  * May use bash or nodejs for general tooling when pragmatic (e.g. better nodejs apis than ruby)
+  * Consider Redis instead of MySQL when performance is important and ephemerality is acceptable (e.g. cache values)
+  * This tenet will probably have more exceptions and edge cases than most. Its not an exhaustive list, its a "highly preferred list".
+    * Be pragmatic, do a good faith attempt to use preferred languages and frameworks, but solve your problems.
+    * E.g. we also use cloudformation (etc etc) which could be argued to be kind of like a programming language, but is **totally** not a problem.
+      * OTOH, you don't get to "use Haskell because everything else should be considered lame and fundamentally broken" 😉.
+
+</details>
+
+### Tenet 5: “The Rails Way” (vs. overfitting to today’s needs)
```
Comment (bethanyaconnor) [inline ARCHITECTURE.md:133]: Is there any good documentation/articles/blog posts that you know of to explain "The Rails Way". Rails is a very opinionated framework, but folks who aren't working in Rails often don't always know what those opinions are.

File: ARCHITECTURE.md (hunk 2 above)
Comment (snickell) [reply ARCHITECTURE.md]: I inherited that line from the previous doc, and also agree that I'd rather write it more simply as supporting caching.

File: ARCHITECTURE.md (hunk 3 above)
Comment (snickell) [reply ARCHITECTURE.md:218]: I think an /ARCHITECTURE.md doc will be at permanent "risk" for somebody deciding to document, well, our architecture there. I'm not trying to encourage a dumping ground here (although personally I would prefer we actually had normal ARCHITECTURE.md material here vs google drive), but I'd rather have a dumping ground below the line than not "defensively" anticipate a possible way this goes off the rails.

File: ARCHITECTURE.md (hunk 4 above)
Comment (snickell) [reply ARCHITECTURE.md:133]: This helps me refine my thinking, and you're right that its pretty broad and abstract. I think addressing it on the level of specific opinions is maybe a losing battle, because part of the point is that there are MANY of them.

The ruby on rails "doctrine" page gives some flavor: https://rubyonrails.org/doctrine

I think that its often quite clear what the opinionated Rails way of doing something is when you try to look on the internet (or ask chatgpt) how to do it. I think part of the memo is "think twice before ignoring that and doing it a way that you learned from django". That is, if you're in the category of people who hasn't learned Rails yet, you also should be extra cautious about "fighting rails". What I think isn't clear coming from frameworks like Django (which is also great, just different) is that in Rails the default has been carefully thought about to be the default, as opposed to Django world where its one of a dozen ways to do things and "flexibility" is valued above "good defaults".

How do I frame this? :-P

File: ARCHITECTURE.md (hunk 2 above)
Comment (snickell) [reply ARCHITECTURE.md]: changed

File: ARCHITECTURE.md (hunk 4 above)
Comment (snickell) [reply ARCHITECTURE.md:133]: The Rails documentation guides toward the conventional way, so do most rails tutorials, so does stack overflow.

> Prefer the conventional Rails way to accomplish a task: it might be better to be 100% conventional than get a 10% gain by deviating from convention. This is not a “don’t do it”, more like a “think twice if you really need to before doing it”.

I think the real memo is that there IS a "rails way" for almost everything, and to learn that, you'll have to search. This is opposed to lots of web frameworks where the goal is almost to not have their be a, say, "django way" (the django way is "however you want to make it work").

Comment (davidsbailey) [reply ARCHITECTURE.md:133]: 1. for people who are completely new to rails, I would recommend a course of some kind
2. for people who are just trying to find the rails way to do a particular thing, I would recommend looking through the guides: https://guides.rubyonrails.org/ (or search for "rails guide" plus whatever you are trying to do)

**Review by simonguest — APPROVED** [review]
Thanks for updating the tenets doc and moving to our repo, @snickell. I like the idea of also using this as a forcing function for creating/expanding our investment into an architecture.md file.

This change also includes a change to dashboard/schema.rb - unsure whether this was intentional or a hanging file from another branch?

File: ARCHITECTURE.md (hunk 4 above)
Comment (snickell) [reply ARCHITECTURE.md:133]: I like https://guides.rubyonrails.org/ as a rec, I expanded it a little into this new paragraph:

> General thoughts on the Rails approach can be found at https://rubyonrails.org/doctrine with more specific advice for "how do to X the rails way" at https://guides.rubyonrails.org/. If you're not familiar with Rails, a course or a book might be helpful.

**Review by ebeastlake — APPROVED** [review]

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+General thoughts on the Rails approach can be found at https://rubyonrails.org/doctrine with more specific advice for "how do to X the rails way" at https://guides.rubyonrails.org/. If you're not familiar with Rails, a course or a book might be helpful.
+
+<details>
+  <summary>Why?</summary>
+
+  * Observing the codebase over a decade of dev: sometimes you can get a gain this year by customizing Rails to work a different way that’s a little better for us, but over time non-conventional code tends to become a maintenance burden.  
+  * Often the “10% better” non-conventional code is exactly what prevents us from upgrading to the “50% better” official Rails solution to the problem that comes out 2 years later.  
+  * By and large we have pretty conventionally Rails-shaped problems, which means Rails tends to solve our problems better than we would if we wait a little longer.  
+  * Rails has been a great long-term fit to our problem space and we benefit from continuing to get upgrades as Rails has improved. Doing it “their way” makes upgrades happen semi-automatically.
```
Comment (davidsbailey) [inline ARCHITECTURE.md:145]: there are some ways we may want to deviate from the rails way due to the size of our application:
* we may want to modularize our rails monolith into separate groups of business logic which own their own data, using rails engines (see [Domain Boundaries - A Starting Point](https://docs.google.com/document/d/1WffY2sr3f-dJQwqj98fhCQhqXhf3bmiaehKlpw21BBc/edit) (2021)). 
* rails steers us toward routes which are tied to models, however we should consider that routes which represent higher-level abstractions (e.g. teacher dashboard) should not be associated with individual models. 

would an exception like this make sense to add here:
* there may be some ways where deviating from the rails way makes sense due to the large size of our application, such as by modularizing our rails app via rails engines or by decoupling rails routes from being associated with individual models.

my thinking is that the emphasis on our application size gives a more general sense of when we might consider deviating, rather than listing the more specific bullets as the exceptions themselves.

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+  * This tenet will probably have more exceptions and edge cases than most. Its not an exhaustive list, its a "highly preferred list".
+    * Be pragmatic, do a good faith attempt to use preferred languages and frameworks, but solve your problems.
+    * E.g. we also use cloudformation (etc etc) which could be argued to be kind of like a programming language, but is **totally** not a problem.
+      * OTOH, you don't get to "use Haskell because everything else should be considered lame and fundamentally broken" 😉.
+
+</details>
+
+### Tenet 5: “The Rails Way” (vs. overfitting to today’s needs)
+
+Prefer the conventional Rails way to accomplish a task: it might be better to be 100% conventional than get a 10% gain by deviating from convention. This is not a “don’t do it”, more like a “think twice if you really need to before doing it”.
```
Comment (davidsbailey) [inline ARCHITECTURE.md:135]: as a litmus test, should someone be able to use this tenet to support PR feedback saying that a new ajax route should be added to regular controllers like controllers/sections_controller, as opposed to our current alternative of using controllers/api/v1/sections_controller? I haven't done an accurate count but I think we use api/v1 about 65% of the time today. 

For example, if the following controller action(s) were being added today, should it be moved to app/controllers? https://github.com/code-dot-org/code-dot-org/blob/f1492ffb32e49b95a61fdc60d19dde7c415132a6/dashboard/app/controllers/api/v1/amazon_future_engineer_controller.rb#L13

Comment (davidsbailey) [reply ARCHITECTURE.md:135]: meta-comment: of all the tenets, i'm concerned that this one leave the most ambiguity around when to follow it vs follow the most common practice in our codebase.

File: ARCHITECTURE.md (hunk 5 above)
Comment (snickell) [reply ARCHITECTURE.md:145]: I like the observation that many deviations from default rails will be driven by the relative size of our app. That said, we're linking to https://guides.rubyonrails.org/ as a first blush "rails way" and we have: https://guides.rubyonrails.org/engines.html, so I'm not sure I'd think of engines as deviating from the rails way?

File: ARCHITECTURE.md (hunk 6 above)
Comment (snickell) [reply ARCHITECTURE.md:135]: I agree this tenet is the most ambiguous, but that's also why its powerful (it tracks lots of specific upstream decisions going forward without having to specify them today, its sort of an `include RailsTenets` directive in the tenets).

That said I was hoping this language makes it clear that unlike many of the tenets (which are SHOULD or MUST), the Rails Way tenet is "PREFER the conventional Rails way".

And additionally is gentled by saying explicitly: "This is not a “don’t do it”, more like a “think twice if you really need to before doing it”.

Comment (snickell) [reply ARCHITECTURE.md:135]: I think the creation of the v1 system itself is a **great** example of "why did we do this weird semi-pointless deviation from rails instead of sticking with rails stock?!?". I wish we could send the tenet back in time to when the v1 routing was added and "think twice" before doing it: what was the real gain here?

But now that we already have some things using this pattern and some not....

> as a litmus test, should someone be able to use this tenet to support PR feedback saying that a new ajax route should be added to regular controllers like controllers/sections_controller

imo probably yes, but at least applying the "not a hard no to deviate from rails way, but think twice" approach advised by the tenet seems wise. It can't make decisions for you, but it can suggest where you might stare twice and ask "What's the advantage of deviating here?"

> should it have been used to support long-ago PR feedback saying "Maybe we shouldn't add this weird v1 api routing system in the first place?"

imo emphatically YES: the creation of our v1 routing "convention" (if you can call it that) is a great example where imo deviating from stock rails proved (afaik) to be weird and pointless, sorta following a fad at the time, and has created future questions like "should someone be able to use this tenet to support PR feedback saying that a new ajax route" due to it being different from stock rails.

Being "Rails-y" takes self education, but in general following Rails has actually been a great technical filter for the last 20 years to avoid adopting technical fads that weren't as important as they seemed in their hottest year.

File: ARCHITECTURE.md (hunk 5 above)
Comment (snickell) [reply ARCHITECTURE.md:145]: I've added this exception:
```
<details>
  <summary>Exceptions</summary>

    * Sometimes deviating from stock Rails makes sense due to the large size of our application:
      * e.g. modularizing our rails app via [Rails engines](https://guides.rubyonrails.org/engines.html).
      * e.g. decoupling rails routes from being associated with individual models.

</details>
```

**Review by bethanyaconnor — APPROVED** [review]

File: ARCHITECTURE.md (hunk 6 above)
Comment (davidsbailey) [reply ARCHITECTURE.md:135]: Thank you, the emphasis on PREFER is helpful.

FWIW there was a real gain for using `api/v1` routes initially -- I believe it was used to simplify the logic for making those routes available to pegasus. However, I agree that api/v1 route prefix no longer appears to serve any purpose. Furthermore, it frequently causes headaches when similar routes are split between multiple controllers with subtly different names and paths.

File: ARCHITECTURE.md (hunk 5 above)
Comment (davidsbailey) [reply ARCHITECTURE.md:145]: oops, nice catch! I think it's still helpful to include, so thanks for adding.

**Review by davidsbailey — APPROVED** [review]

File: ARCHITECTURE.md (hunk 6 above)
Comment (snickell) [reply ARCHITECTURE.md:135]: Oh interesting, always nice to learn history for codebases, really helps understanding

**Review by cat5inthecradle — APPROVED** [review]
